src/metabase/api/user.clj

```diff
@@ -177,27 +198,42 @@
   Takes `query` for filtering on first name, last name, email.
   Also takes `group_id`, which filters on group id."
   [status query group_id include_deactivated]
-  {status                 (s/maybe s/Str)
-   query                  (s/maybe s/Str)
-   group_id               (s/maybe su/IntGreaterThanZero)
-   include_deactivated    (s/maybe su/BooleanString)}
-  (when (or status include_deactivated)
-    (validation/check-group-manager))
+    {status              (s/maybe s/Str)
+     query               (s/maybe s/Str)
+     group_id            (s/maybe su/IntGreaterThanZero)
+     include_deactivated (s/maybe su/BooleanString)}
+  (validation/check-group-manager)
+  (when group_id
+    (check-manager-of-group group_id))
   (let [include_deactivated (Boolean/parseBoolean include_deactivated)
-        clauses             (user-clauses status query (if (some? group_id) [group_id] nil) include_deactivated)]
-    {:data   (cond-> (t2/select
-                      (vec (cons User (user-visible-columns)))
-                      (cond-> clauses
-                        (some? group_id) (sql.helpers/order-by [:core_user.is_superuser :desc] [:is_group_manager :desc])
-                        true             (sql.helpers/order-by [:%lower.first_name :asc] [:%lower.last_name :asc])))
-               ;; For admins also include the IDs of Users' Personal Collections
-               api/*is-superuser?*
-               (t2/hydrate :personal_collection_id)
-
-               (or api/*is-superuser?*
-                   api/*is-group-manager?*)
-               (t2/hydrate :group_ids))
-     :total  (t2/count User (filter-clauses-without-paging clauses))
+        manager-group-ids   (set (group-ids-for-manager api/*current-user-id*))
+        group-id-clause     (cond
+                              ;; superuser can see anything, so group_id clause is ok
+                              (and api/*is-superuser?* group_id) [group_id]
+                              ;; otherwise, if the user is a group manager, only show them users in the groups they manage
+                              api/*is-group-manager?*            (vec manager-group-ids))
```

Comment:
Think this logic is still slightly off. What if the user is a group manager _and_ `group_id` is passed in? I think if `group_id` is passed in we want add the filter unconditionally at this point (since we've already very that the user is either an admin or manager of the group), and if it's not passed in, we only want to add a filter for group managers.

```suggestion
        group-id-clause     (cond
                              ;; superuser can see anything, so group_id clause is ok
                              group_id [group_id]
                              ;; otherwise, if the user is a group manager, only show them users in the groups they manage
                              api/*is-group-manager?* (vec manager-group-ids))
```